src/targets/gpu/kernels/include/migraphx/kernels/softmax.hpp

```diff
@@ -33,11 +33,11 @@ template <index_int Axis, class Input, class Output>
 __device__ void softmax(Input input, Output output)
 {
     reduce::block::run<reduce::with_axis<Input, Axis>>([&](auto, auto r) {
-        auto batch_max = r.reduce(op::max{}, lowest{}, op::id{})(input);
-        auto batch_sum =
-            r.reduce(op::sum{}, 0, [&](auto x) { return migraphx::exp(x - batch_max); })(input);
-        r.inner([&](auto& y, auto x) { y = migraphx::exp(x - batch_max) / batch_sum; })(output,
-                                                                                        input);
+        const auto c   = vec_at(r.slice(input)[0], 0);
+        auto batch_sum = r.reduce(op::sum{}, 0, [&](auto x) {
+            return migraphx::convert<float>(migraphx::exp(x - c));
+        })(input);
+        r.inner([&](auto& y, auto x) { y = migraphx::exp(x - c) / batch_sum; })(output, input);
```

Comment:
I think we should still being using the max in the calculation, but I've seen an alternative to still save us from blowing up our exp() and prevents overflow and solve the correctness issue with Naans

calculate log(output) instead and then convert it back

This gives us the following:

```
//since
batch_sum = sum(exp(input))
max   = max(input)

batch_sum_sub_max = exp(input - max)
output = exp(x) / batch_sum =  exp(x-max) / batch_sum_sub_max
//take natural log of the sum
ln(output) =  (x - max) - log(batch_sum_sub_max)
//thus
output = exp( x - max - log(batch_sum_sub_max))
```

Refs:
https://stackoverflow.com/questions/9906136/implementation-of-a-softmax-activation-function-for-neural-networks
https://lingpipe-blog.com/2009/06/25/log-sum-of-exponentials/

The issue here now is will we have a speedup then without the max? Unless we have a fast way of doing the log() operator,